modules/xfeatures2d/src/msd.cpp

```diff
@@ -0,0 +1,791 @@
+                    pt.y = m_search_area_radius * sin(deltaAngle);
+
+                    orientPoints.push_back(pt);
+
+                    deltaAngle += step;
+                }
+            }
+
+            for (int r = 0; r < m_cur_n_scales; r++)
+            {
+                int cW = m_scaleSpace[r].cols;
+                int cH = m_scaleSpace[r].rows;
+
+                for (int j = border; j < cH - border; j++)
+                {
+                    for (int i = border; i < cW - border; i++)
+                    {
+                        if (saliency[r][j * cW + i] <= m_th_saliency)
+                            continue;
+
+                        if (m_mask.rows > 0)
+                        {
+                            int j_full = cvRound(j * std::pow(m_scale_factor, r));
+                            int i_full = cvRound(i * std::pow(m_scale_factor, r));
+                            if (m_mask.at<unsigned char>(j_full, i_full) == 0)
+                                continue;
+                        }
+
+                        bool is_max = true;
+
+                        for (int k = cv::max(0, r - m_nms_scale_radius); k <= cv::min(m_cur_n_scales - 1, r + m_nms_scale_radius); k++)
+                        {
+                            if (k != r)
+                            {
+                                int j_sc = cvRound(j * std::pow(m_scale_factor, r - k));
+                                int i_sc = cvRound(i * std::pow(m_scale_factor, r - k));
+
+                                if (saliency[r][j * cW + i] < saliency[k][j_sc * cW + i_sc])
+                                {
+                                    is_max = false;
+                                    break;
+                                }
+                            }
+                        }
+
+                        for (int v = cv::max(border, j - m_nms_radius); v <= cv::min(cH - border - 1, j + m_nms_radius); v++)
+                        {
+                            for (int u = cv::max(border, i - m_nms_radius); u <= cv::min(cW - border - 1, i + m_nms_radius); u++)
+                            {
+                                if (saliency[r][j * cW + i] < saliency[r][v * cW + u])
+                                {
+                                    is_max = false;
+                                    break;
+                                }
+                            }
+
+                            if (!is_max)
+                                break;
+                        }
+
+                        if (is_max)
+                        {
+                            bool resInt = rescalePoint(i, j, r, saliency, kp_temp.pt);
+                            if (!resInt)
+                                continue;
+
+
+                            if (m_mask.rows > 0)
+                            {
+                                if (m_mask.at<unsigned char>(kp_temp.pt.y, kp_temp.pt.x) == 0)
```

Comment:
You have a possible loss of accuracy due to the float and int comparison.
Solve this by making it: `(int)m_mask.at<unsigned char>(kp_temp.pt.y, kp_temp.pt.x) == 0`